### ISSUE TITLE
Using SFTPGo as FTP Server for Lexmark Printer fails

### ISSUE THREAD

**Issue by markusressel** [drakkan/sftpgo#1082]
Hey there,

I am trying to use your docker container to get my Lexmark X464 printer to upload scanned files to a SFTPGo FTP server.

Sadly all my attempts have failed so far, as I am stuck at this error from the printer:

```
504 Unsupported transfer type
```

Connecting from a shell using `ftp` and issuing `ls` gives me this:
```
╰─λ ftp paperless@192.168.2.138
Connected to 192.168.2.138.
220 SFTPGo 2.4.2 ready
331 OK
Password: 
230 Password ok, continue
Remote system type is UNIX.
Using binary mode to transfer files.
ftp> ls
501 Your request does not meet the configured security requirements
500 Unknown command "LPRT"
ftp: bind: Address already in use
```

Trying to `put` a file gives me this:
```
╰─λ ftp paperless@192.168.2.138                                                                                                                                                                                                            0 (7.819s) < 01:05:14
Connected to 192.168.2.138.
220 SFTPGo 2.4.2 ready
331 OK
Password: 
230 Password ok, continue
Remote system type is UNIX.
Using binary mode to transfer files.
ftp> put README.md
501 Your request does not meet the configured security requirements
500 Unknown command "LPRT"
ftp: bind: Address already in use
ftp> 
```

However, connecting from FileZilla works fine. I can even upload, download and delete files without issues.

I have since tried three other docker containers to get this running, but all of them have their own set of issues... SFTPGo still seems like the best option, if we can get this to work.

Any help would be much appreviated.

###  My SFTPGo Config (env only):
```
  SFTPGO_CREATE_DEFAULT_ADMIN: "true"
  SFTPGO_DEFAULT_ADMIN_USERNAME: paperless
  SFTPGO_DEFAULT_ADMIN_PASSWORD: paperless
  SFTPGO_FTPD__DEBUG: "true"
  SFTPGO_FTPD__BINDINGS__0__PORT: "2121"
  SFTPGO_FTPD__BINDINGS__0__FORCE_PASSIVE_IP: "192.168.2.138"
  SFTPGO_FTPD__PASSIVE_PORT_RANGE__START: "50000"
  SFTPGO_FTPD__PASSIVE_PORT_RANGE__END:   "50009"
```

**Comment by drakkan** [comment]
Hi,

you haven't shared the SFTPGo logs. If you connect from the shell try to switch to passive mode like this

```
ftp a@127.0.0.1 2121
Connected to 127.0.0.1.
220 SFTPGo 2.4.2-dev ready
331 OK
Password: 
230 Password ok, continue
Remote system type is UNIX.
Using binary mode to transfer files.
ftp> passive
Passive mode on.
ftp> ls
227 Entering Passive Mode (127,0,0,1,195,145)
150 Using transfer connection
226 Closing transfer connection
```

FileZilla and other clients generally use passive mode. 
Does the printer connect using active or passive mode? Anyway, if you share SFTPGo logs with FTP debugging enabled it should be clear what's going on.

You can also relax the security requirements for active connections by adding this env var

```
SFTPGO_FTPD__BINDINGS__0__ACTIVE_CONNECTIONS_SECURITY: "1"
```

**Comment by drakkan** [comment]
This error

> 504 Unsupported transfer type

is returned if the printer sends an unsupported transfer type, we support binary and ascii mode. It would be interesting to understand what the printer sends, probably an alias for one of these modes. The debug logs should contain the transfer type. Thanks

**Comment by markusressel** [comment]
Thx for the quick answer, sadly I didn't find much of interest in the logs so far, this is the complete output (only the last 4-5 lines are printed during the printer's operation):

<details>
<summary>Logs</summary>

```
{"level":"info","time":"2022-12-03T11:10:54.914","sender":"common","message":"scheduled overquota transfers check, schedule \"@every 1m0s\""}
{"level":"info","time":"2022-12-03T11:10:54.914","sender":"common","message":"scheduled idle connections check, schedule \"@every 3m0s\""}
{"level":"info","time":"2022-12-03T11:10:54.914","sender":"common","message":"using memory transfer checker"}
{"level":"info","time":"2022-12-03T11:10:54.914","sender":"kms","message":"secret provider registered for scheme: \"builtin\", encrypted status: \"AES-256-GCM\""}
{"level":"info","time":"2022-12-03T11:10:54.914","sender":"kms","message":"secret provider registered for scheme: \"local\", encrypted status: \"Secretbox\""}
{"level":"debug","time":"2022-12-03T11:10:54.915","sender":"plugins","message":"initialize"}
{"level":"debug","time":"2022-12-03T11:10:54.915","sender":"smtp","message":"configuration disabled, email capabilities will not be available"}
{"level":"debug","time":"2022-12-03T11:10:54.915","sender":"dataprovider","message":"absolute backup path \"/srv/sftpgo/backups\""}
{"level":"debug","time":"2022-12-03T11:10:54.915","sender":"dataprovider_sqlite","message":"sqlite database handle created, connection string: \"file:sftpgo.db?cache=shared&_foreign_keys=1\""}
{"level":"debug","time":"2022-12-03T11:10:54.963","sender":"dataprovider_sqlite","message":"sql database is up to date, current version: 23"}
{"level":"debug","time":"2022-12-03T11:10:54.964","sender":"eventmanager","message":"loading updated rules"}
{"level":"debug","time":"2022-12-03T11:10:54.964","sender":"dataprovider_sqlite","message":"delayed quota update loop started, wait time: 0s"}
{"level":"debug","time":"2022-12-03T11:10:54.964","sender":"dataprovider_sqlite","message":"delayed quota update loop ended, wait time: 0s"}
{"level":"debug","time":"2022-12-03T11:10:54.965","sender":"eventmanager","message":"recently updated event rules loaded: 0"}
{"level":"debug","time":"2022-12-03T11:10:54.965","sender":"eventmanager","message":"event rules updated, fs events: 0, provider events: 0, schedules: 0, ip blocked events: 0, certificate events: 0"}
{"level":"info","time":"2022-12-03T11:10:54.965","sender":"acme","message":"no domains configured, acme disabled"}
{"level":"info","time":"2022-12-03T11:10:54.966","sender":"service","message":"WebDAV server not started, disabled in config file"}
{"level":"info","time":"2022-12-03T11:10:54.966","sender":"service","message":"telemetry server not started, disabled in config file"}
{"level":"info","time":"2022-12-03T11:10:54.966","sender":"service","message":"initializing SFTP server with config {Banner:SFTPGo_2.4.2 Bindings:[{Address: Port:2022 ApplyProxyConfig:true}] MaxAuthTries:0 HostKeys:[] HostCertificates:[] HostKeyAlgorithms:[] Moduli:[] KexAlgorithms:[] Ciphers:[] MACs:[] TrustedUserCAKeys:[] RevokedUserCertsFile: LoginBannerFile: EnabledSSHCommands:[md5sum sha1sum sha256sum cd pwd scp] KeyboardInteractiveAuthentication:false KeyboardInteractiveHook: PasswordAuthentication:true FolderPrefix: certChecker:<nil> parsedUserCAKeys:[]}"}
{"level":"info","time":"2022-12-03T11:10:54.966","sender":"sftpd","message":"Loading private host key \"id_rsa\""}
{"level":"info","time":"2022-12-03T11:10:54.971","sender":"ftpd","message":"initializing FTP server with config {Bindings:[{Address: Port:2121 ApplyProxyConfig:true TLSMode:0 CertificateFile: CertificateKeyFile: MinTLSVersion:12 ForcePassiveIP:192.168.2.138 PassiveIPOverrides:[] ClientAuthType:0 TLSCipherSuites:[] PassiveConnectionsSecurity:0 ActiveConnectionsSecurity:1 Debug:true ciphers:[]}] Banner:SFTPGo 2.4.2 ready BannerFile: CertificateFile: CertificateKeyFile: CACertificates:[] CARevocationLists:[] ActiveTransfersPortNon20:true DisableActiveMode:false EnableSite:false HASHSupport:0 CombineSupport:0 PassivePortRange:{Start:50000 End:50009}}"}
{"level":"info","time":"2022-12-03T11:10:54.971","sender":"ftpd","message":"starting FTP serving, binding: :2121"}
{"level":"info","time":"2022-12-03T11:10:54.972","sender":"ftpserverlib","server_id":"FTP_0","address":"[::]:2121","message":"Listening..."}
{"level":"info","time":"2022-12-03T11:10:54.972","sender":"ftpserverlib","server_id":"FTP_0","message":"Starting..."}
{"level":"info","time":"2022-12-03T11:10:54.969","sender":"httpd","message":"initializing HTTP server with config {Bindings:[{Address: Port:8080 EnableWebAdmin:true EnableWebClient:true EnableRESTAPI:true EnabledLoginMethods:0 EnableHTTPS:false CertificateFile: CertificateKeyFile: MinTLSVersion:12 ClientAuthType:0 TLSCipherSuites:[] ProxyAllowed:[] ClientIPProxyHeader: ClientIPHeaderDepth:0 HideLoginURL:0 RenderOpenAPI:true WebClientIntegrations:[] OIDC:{ClientID: ClientSecret: ConfigURL: RedirectBaseURL: UsernameField: RoleField: ImplicitRoles:false Scopes:[openid profile email] CustomFields:[] InsecureSkipSignatureCheck:false Debug:false provider:<nil> verifier:<nil> providerLogoutURL: oauth2Config:<nil>} Security:{Enabled:false AllowedHosts:[] AllowedHostsAreRegex:false HostsProxyHeaders:[] HTTPSRedirect:false HTTPSHost: HTTPSProxyHeaders:[] STSSeconds:0 STSIncludeSubdomains:false STSPreload:false ContentTypeNosniff:false ContentSecurityPolicy: PermissionsPolicy: CrossOriginOpenerPolicy: ExpectCTHeader: proxyHeaders:[]} Branding:{WebAdmin:{Name: ShortName: LogoPath: LoginImagePath: FaviconPath: DisclaimerName: DisclaimerPath: DefaultCSS: ExtraCSS:[]} WebClient:{Name: ShortName: LogoPath: LoginImagePath: FaviconPath: DisclaimerName: DisclaimerPath: DefaultCSS: ExtraCSS:[]}} allowHeadersFrom:[]}] TemplatesPath:templates StaticFilesPath:static OpenAPIPath:openapi WebRoot: CertificateFile: CertificateKeyFile: CACertificates:[] CARevocationLists:[] SigningPassphrase: TokenValidation:0 MaxUploadFileSize:1048576000 Cors:{AllowedOrigins:[] AllowedMethods:[] AllowedHeaders:[] ExposedHeaders:[] AllowCredentials:false Enabled:false MaxAge:0 OptionsPassthrough:false OptionsSuccessStatus:0 AllowPrivateNetwork:false} Setup:{InstallationCode: InstallationCodeHint:Installation code} HideSupportLink:false}"}
{"level":"info","time":"2022-12-03T11:10:54.972","sender":"httpd","message":"using memory reset code manager"}
{"level":"info","time":"2022-12-03T11:10:54.972","sender":"httpd","message":"using memory OIDC manager"}
{"level":"debug","time":"2022-12-03T11:10:54.973","sender":"util","message":"found share data path for name \"static\": \"/usr/share/sftpgo/static\""}
{"level":"debug","time":"2022-12-03T11:10:54.973","sender":"util","message":"found share data path for name \"templates\": \"/usr/share/sftpgo/templates\""}
{"level":"debug","time":"2022-12-03T11:10:54.973","sender":"util","message":"found share data path for name \"openapi\": \"/usr/share/sftpgo/openapi\""}
{"level":"info","time":"2022-12-03T11:10:54.985","sender":"sftpd","message":"Host key \"id_rsa\" loaded, type \"ssh-rsa\", fingerprint \"SHA256:8/ewQiK/AlCQ3Lm6nF54V47863sIYWxueVxVRsomD8o\""}
{"level":"info","time":"2022-12-03T11:10:54.985","sender":"sftpd","message":"Loading private host key \"id_ecdsa\""}
{"level":"info","time":"2022-12-03T11:10:55.002","sender":"sftpd","message":"Host key \"id_ecdsa\" loaded, type \"ecdsa-sha2-nistp256\", fingerprint \"SHA256:HBeqDrkRnlqaFmCHW25XsG6G/awWMrUPda+WwE6SElY\""}
{"level":"info","time":"2022-12-03T11:10:55.002","sender":"sftpd","message":"Loading private host key \"id_ed25519\""}
{"level":"info","time":"2022-12-03T11:10:55.022","sender":"sftpd","message":"Host key \"id_ed25519\" loaded, type \"ssh-ed25519\", fingerprint \"SHA256:lEIBcgTqho9OgV4+lB0OpyqrUQoGlbA1sH9+zndSDaA\""}
{"level":"debug","time":"2022-12-03T11:10:55.022","sender":"sftpd","message":"enabled SSH commands [md5sum sha1sum sha256sum cd pwd scp]"}
{"level":"info","time":"2022-12-03T11:10:55.023","sender":"sftpd","message":"server listener registered, address: [::]:2022"}
{"level":"info","time":"2022-12-03T11:10:55.087","sender":"httpd","message":"server listener registered, address: [::]:8080 TLS enabled: false"}
{"level":"debug","time":"2022-12-03T11:12:58.545","sender":"ftpserverlib","server_id":"FTP_0","clientId":"1","clientIp":"10.4.0.2:33383","message":"Client connected"}
{"level":"debug","time":"2022-12-03T11:12:58.546","sender":"FTP","connection_id":"FTP_0_1","message":"connection added, local address \"10.42.1.106:2121\", remote address \"10.4.0.2:33383\", num open connections: 1"}
{"level":"debug","time":"2022-12-03T11:12:58.547","sender":"ftpserverlib","server_id":"FTP_0","clientId":"1","line":"220 SFTPGo 2.4.2 ready","message":"Sending answer"}
{"level":"debug","time":"2022-12-03T11:12:58.571","sender":"ftpserverlib","server_id":"FTP_0","clientId":"1","line":"USER paperless","message":"Received line"}
{"level":"debug","time":"2022-12-03T11:12:58.572","sender":"ftpserverlib","server_id":"FTP_0","clientId":"1","line":"331 OK","message":"Sending answer"}
{"level":"debug","time":"2022-12-03T11:12:58.580","sender":"ftpserverlib","server_id":"FTP_0","clientId":"1","line":"PASS paperless","message":"Received line"}
{"level":"debug","time":"2022-12-03T11:12:58.656","sender":"common","connection_id":"FTP_0_1","message":"connection swapped, close fs error: <nil>"}
{"level":"info","time":"2022-12-03T11:12:58.656","sender":"FTP","connection_id":"FTP_0_1","message":"User \"paperless\" logged in with \"password\" from ip \"10.4.0.2\""}
{"level":"debug","time":"2022-12-03T11:12:58.688","sender":"dataprovider_sqlite","message":"last login updated for user \"paperless\""}
{"level":"debug","time":"2022-12-03T11:12:58.688","sender":"ftpserverlib","server_id":"FTP_0","clientId":"1","line":"230 Password ok, continue","message":"Sending answer"}
{"level":"debug","time":"2022-12-03T11:12:58.733","sender":"ftpserverlib","server_id":"FTP_0","clientId":"1","line":"TYPE i","message":"Received line"}
{"level":"debug","time":"2022-12-03T11:12:58.733","sender":"ftpserverlib","server_id":"FTP_0","clientId":"1","line":"504 Unsupported transfer type","message":"Sending answer"}
{"level":"debug","time":"2022-12-03T11:12:59.184","sender":"ftpserverlib","server_id":"FTP_0","clientId":"1","line":"QUIT","message":"Received line"}
{"level":"debug","time":"2022-12-03T11:12:59.184","sender":"ftpserverlib","server_id":"FTP_0","clientId":"1","line":"221 Goodbye","message":"Sending answer"}
{"level":"debug","time":"2022-12-03T11:12:59.184","sender":"ftpserverlib","server_id":"FTP_0","clientId":"1","clean":"true","message":"Client disconnected"}
{"level":"debug","time":"2022-12-03T11:12:59.185","sender":"FTP","connection_id":"FTP_0_1","message":"connection removed, local address \"10.42.1.106:2121\", remote address \"10.4.0.2:33383\" close fs error: <nil>, num open connections: 0"}
{"level":"debug","time":"2022-12-03T11:12:59.185","sender":"ftpserverlib","server_id":"FTP_0","clientId":"1","clientIp":"10.4.0.2:33383","message":"Client disconnected"}
{"level":"debug","time":"2022-12-03T11:12:59.185","sender":"ftpserverlib","server_id":"FTP_0","clientId":"1","err":"close tcp 10.42.1.106:2121->10.4.0.2:33383: use of closed network connection","message":"Problem closing control connection"}

```

</details>

**Comment by markusressel** [comment]
I can't really tell what mode the printer is using, but based on other error messages I got during the whole setup process, some of them I could only reproduce while NOT using the passive mode. Therefore I guess that the printer is using the active mode, but I am not sure how to figure this out definitely.

**Comment by drakkan** [comment]
This is the issue `TYPE i`, give me some hours to prepare a patch for the ftp library we use

**Comment by drakkan** [comment]
Please give a try to one of the latest edge based docker images. The parameter for the `TYPE` command should be now handled in a [case-insensitive manner](https://github.com/drakkan/ftpserverlib/commit/ba73c775a9fdb5e51ce39723c52e91f8484ef4cb). Let's see the next error :smile: 

To confirm that you are using the correct version you should see this log `starting SFTPGo 2.4.2-dev-2ea43647`. Thanks

**Comment by markusressel** [comment]
> Let's see the next error :smile: 

Well we cannot do that because... it already works! :partying_face: :tada:
Scanned documents are uploaded without issue now. :muscle:
Thx for the quick update :heart:

No need to rush (its working fine using the edge container), but when this moves to an "officially versioned" release, it would be great if you could notify me so I can switch the container image.

**Comment by drakkan** [comment]
This fix will be included in v2.4.3. If you don't want to use the development version (it is generally as stable as the released versions), you can use a 2.4.x docker image. 2.4.x is the "stable" branch + fixes, v2.4.3 will be tagged from this branch

https://github.com/drakkan/sftpgo/commits/2.4.x
